beacons/src/commonMain/kotlin/Beacons.kt

```diff
@@ -17,45 +17,110 @@
 
 package com.splendo.kaluga.bluetooth.beacons
 
+import co.touchlab.stately.collections.IsoMutableMap
+import co.touchlab.stately.concurrency.AtomicReference
 import com.splendo.kaluga.base.utils.Date
 import com.splendo.kaluga.bluetooth.BluetoothService
 import com.splendo.kaluga.bluetooth.device.Device
-import com.splendo.kaluga.bluetooth.device.Identifier
+import com.splendo.kaluga.logging.debug
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.flow.firstOrNull
 import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
 
 class Beacons(
     private val bluetooth: BluetoothService,
-    private val timeoutMs: Int = 10_000
+    private val timeoutMs: Long = 10_000,
 ) {
 
-    fun startMonitoring() = bluetooth.startScanning()
-    fun stopMonitoring() = bluetooth.stopScanning()
+    private companion object { const val TAG = "Beacons" }
+
+    private val cache = IsoMutableMap<String, Pair<BeaconInfo, Job>>()
+    private val cacheJob = Job()
+    private val coroutineScope = CoroutineScope(Dispatchers.Default + cacheJob)
+
+    private val _monitoringJob = AtomicReference<Job?>(null)
+    private var monitoringJob: Job?
```

Comment:
If you use `AtomicReference` don't forget to set this as null in stopMonitoring otherwise [memory leak](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.native.concurrent/-atomic-reference/) could happen. If it makes things easier and sense you could try to use it as a StateFlow..

Reply:
Good point